lib/remap-functions/src/parse_common_log.rs

```diff
@@ -0,0 +1,242 @@
+        let message = String::from_utf8_lossy(&bytes);
+
+        let mut log: BTreeMap<String, Value> = BTreeMap::new();
+
+        let captures = REGEX_COMMON_LOG
+            .captures(&message)
+            .ok_or("failed parsing common log line")?;
+
+        if let Some(host) = captures.name("host").map(|capture| capture.as_str()) {
+            log.insert("host".into(), Value::Bytes(host.to_owned().into()));
+        }
+
+        if let Some(identity) = captures.name("identity").map(|capture| capture.as_str()) {
+            log.insert("identity".into(), Value::Bytes(identity.to_owned().into()));
+        }
+
+        if let Some(user) = captures.name("user").map(|capture| capture.as_str()) {
+            log.insert("user".into(), Value::Bytes(user.to_owned().into()));
+        }
+
+        if let Some(timestamp) = captures.name("timestamp").map(|capture| capture.as_str()) {
+            log.insert(
+                "timestamp".into(),
+                Value::Timestamp(
+                    DateTime::parse_from_str(timestamp, &self.timestamp_format)
+                        .map_err(|error| {
+                            format!(
+                                r#"failed parsing timestamp {} using format {}: {}"#,
+                                timestamp, self.timestamp_format, error
+                            )
+                        })?
+                        .into(),
+                ),
+            );
+        }
+
+        if let Some(message) = captures.name("message").map(|capture| capture.as_str()) {
+            log.insert("message".into(), Value::Bytes(message.to_owned().into()));
+        }
+
+        if let Some(method) = captures.name("method").map(|capture| capture.as_str()) {
+            log.insert("method".into(), Value::Bytes(method.to_owned().into()));
+        }
+
+        if let Some(path) = captures.name("path").map(|capture| capture.as_str()) {
+            log.insert("path".into(), Value::Bytes(path.to_owned().into()));
+        }
+
+        if let Some(protocol) = captures.name("protocol").map(|capture| capture.as_str()) {
+            log.insert("protocol".into(), Value::Bytes(protocol.to_owned().into()));
+        }
+
+        if let Some(status) = captures.name("status").map(|capture| capture.as_str()) {
+            log.insert(
+                "status".into(),
+                Value::Integer(status.parse().map_err(|_| "failed parsing status code")?),
+            );
+        }
+
+        if let Some(size) = captures.name("size").map(|capture| capture.as_str()) {
+            log.insert(
+                "size".into(),
+                Value::Integer(size.parse().map_err(|_| "failed parsing content length")?),
+            );
+        }
+
+        Ok(log.into())
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        self.value
+            .type_def(state)
+            .fallible_unless(value::Kind::Bytes)
+            .with_constraint(value::Kind::Map)
```

Comment:
Have merged the inner type defs PR now, so this needs to specify the field types for the return. Something like:
```rust
            .with_inner_type(inner_type_def())
```
and
```rust
fn inner_type_def() -> Option<InnerTypeDef> {
    Some(inner_type_def! ({
       "host": Kind::Bytes,
       "identity": Kind::Bytes,
       "user": Kind::Bytes,
       "timestamp": Kind::Timestamp,
       "message": Kind::Bytes,
       "method": Kind::Bytes,
       "path": Kind::Bytes,
       "protocol": Kind::Bytes,
       "status": Kind::Integer
       "size": Kind::Integer
    })
}
```

I'm assuming here that all fields must exist. If a field is optional the type would be eg.. `Kind::Bytes | Kind::Null`.